crates/rust-analyzer/src/cli/lsif.rs

```diff
@@ -0,0 +1,146 @@
+            load_out_dirs_from_check: true,
+            with_proc_macro: true,
+            prefill_caches: false,
+        };
+        let path = AbsPathBuf::assert(env::current_dir()?.join(&self.path));
+        let manifest = ProjectManifest::discover_single(&path)?;
+
+        let workspace = ProjectWorkspace::load(manifest, &cargo_config, no_progress)?;
+
+        let (host, vfs, _proc_macro) = load_workspace(workspace, &load_cargo_config)?;
+        let db = host.raw_database();
+        let analysis = host.analysis();
+
+        let work = all_modules(db).into_iter().filter(|module| {
+            let file_id = module.definition_source(db).file_id.original_file(db);
+            let source_root = db.file_source_root(file_id);
+            let source_root = db.source_root(source_root);
+            !source_root.is_library
+        });
+
+        let mut visited_files = FxHashSet::default();
+
+        let mut lsif = LsifManager::default();
+
+        lsif.add(Element::Vertex(Vertex::MetaData {
+            version: String::from("0.5.0"),
+            project_root: lsp_types::Url::from_file_path(path).unwrap(),
+            position_encoding: Encoding::Utf16,
+            tool_info: None,
+        }));
+
+        for module in work {
+            let file_id = module.definition_source(db).file_id.original_file(db);
+            if !visited_files.contains(&file_id) {
+                let path = vfs.file_path(file_id);
+                let path = path.as_path().unwrap();
+                let doc_id = lsif.add(Element::Vertex(Vertex::Document(Document {
+                    language_id: Language::Rust,
+                    uri: lsp_types::Url::from_file_path(path).unwrap(),
+                })));
+                let folds = analysis.folding_ranges(file_id)?;
+                let text = analysis.file_text(file_id)?;
+                let line_index = db.line_index(file_id);
+                let result = folds
+                    .into_iter()
+                    .map(|it| to_proto::folding_range(&*text, &LineIndex::with_default_options(line_index.clone()), false, it))
+                    .collect();
+                let folding_id = lsif.add(Element::Vertex(
+                    Vertex::FoldingRangeResult { result }
+                ));
+                lsif.add(Element::Edge(Edge::FoldingRange(
+                    EdgeData {
+                        in_v: folding_id.into(),
+                        out_v: doc_id.into(),
+                    }
+                )));
+                visited_files.insert(file_id);
+            }
+        }
+        Ok(())
+    }
+}
+
+fn all_modules(db: &dyn HirDatabase) -> Vec<Module> {
+    let mut worklist: Vec<_> =
+        Crate::all(db).into_iter().map(|krate| krate.root_module(db)).collect();
+    let mut modules = Vec::new();
+
+    while let Some(module) = worklist.pop() {
+        modules.push(module);
+        worklist.extend(module.children(db));
+    }
+
+    modules
+}
```

Comment:
One problem here is that this module will contain substantial amounts of logic, and it will be hard to test (everything in `rust-analyzer` crate is more or less untested, because it depends on a lot of outside environment). 

I think we should do this using a pattern we employ for every other LSP feature and implement this in the ide layer, leaving to `rust-analyzer` just a straight-forward translation of internal rust-specific types to external, langauge-independent types. 

Concreterly, I imagine that we want to add an `ide_static_index` crate, which contains something like

```rust
/// A static representation of fully analyzed source code. 
///
/// The intended use-case is powering read-only code browsers and emitting LSIF
pub struct StaticIndex {
    
}

impl StaticIndex {
  pub fn compute(db: &RootDatabase) -> StaticIndex
}
```

Crucially, `ide_static_index` should *not* itself depend on any lsp or lsif types.

the `ide` crate will then re-export that as

```rust
impl Analysis {
    pub fn static_index(&self) -> Result<StaticIndex, Cancelled>
}
```

and in the rust-analyzer crate we'd have `to_proto::lsif` which would contain essentially `impl From<StaticIndex> for LSIF`.

See https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/architecture.md#crateside and https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/architecture.md#cratesrust-analyzer for the motivation on what goes where in terms of crate structure.

Reply:
EDIT: maybe we don't need `ide_static_index` crate just yet and just a submodule in the ide crate will be enough!

Reply:
I added a static_index module inside ide